File: Model/class_Model_LosslessQuad.m
classdef class_Model_LosslessQuad < matlab.mixin.Copyable
    properties
        name    % name of the model
        nx      % number of states
        
        % dynamics at the origin: xdot = c + Lx + phi(x)
        c       % constant dynamics
        L       % linear dynamics
        Ls      % symmetric part of linear dynamics
        La      % asymmetric part of linear dynamics
        Q       % quadratic dynamics
        
        % dynamics at the coordinate shift ydot = d + Ay + phi(y)
        m       % shifted coordinate, default to be 0. 
        d       % constant dynamics
        A       % linear dynamics
        As      % symmetric part of linear dynamics
        Aa      % asymmetric part of linear dynamics
        
        % 
        ode
        dK0 
        ode_shifted
        dKm
    end
    
    methods
        function obj = class_Model_LosslessQuad(name, c, L, Q, m)
            obj.name = name;
            obj.nx = size(c,1);
            
            obj.c = c;
            obj.L = L;
            obj.Ls = 1/2*(L+L');
            obj.La = L - obj.Ls;
            obj.Q = Q;
            
            obj = setup_dyn(obj);
            
            if nargin ~= 5
                m = zeros(obj.nx,1);
            end
            obj = func_ShiftSystem(obj, m);                
        end
        
        function obj = setup_dyn(obj)
            obj.ode = @(t,x) ode_quadraticDyn(obj.c, obj.L, obj.Q, t, x);
            obj.dK0 = @(x) obj.c'*x + x'*obj.Ls*x;
        end
        
        function obj = func_ShiftSystem(obj, m)
            obj.m = m;
            
            % update constant part
            obj.d = zeros(obj.nx,1);
            %     d = sym('d',[nx,1],'real');
            for i = 1:obj.nx
                obj.d(i) = obj.c(i) + obj.L(i,:)*m + m'*obj.Q(:,:,i)*m;
            end
           
            % update linear part
            obj.A = zeros(obj.nx,obj.nx);
            %     A = sym('A',[nx,nx],'real');
            for i = 1:obj.nx
                obj.A(i,:) = obj.L(i,:) + 2*m'*obj.Q(:,:,i);
            end
            
            obj.As = 1/2*(obj.A+obj.A');
            %             % alternative:
            %             obj.As = 1/2*(obj.L+obj.L');
            %             for i = 1:obj.nx
            %                 obj.As = obj.As - m(i)*obj.Q(:,:,i);
            %             end
            obj.Aa = obj.A - obj.As;
            
            % update ode and power function
            obj.ode_shifted = @(t,x) ode_quadraticDyn(obj.d, obj.A, obj.Q, t, x);
            obj.dKm = @(x) obj.d'*x + x'*obj.As*x;
        end
        
        function model_invTime = get_inverseTimeModel(obj)
            cInv = -obj.c;
            LInv = -obj.L;
            QInv = -obj.Q;
            
            model_invTime = class_Model_LosslessQuad(obj.name, ...
                cInv, LInv, QInv, obj.m);
        end
    end
end

File: Model/class_Model_LosslessQuad.py
import numpy as np
import cvxpy as cp
from typing import Callable, Optional, Union
from dataclasses import dataclass, field
import copy

def ode_quadraticDyn(c: np.ndarray, L: np.ndarray, Q: np.ndarray, t: float, x: np.ndarray) -> np.ndarray:
    """Helper function to compute quadratic dynamics."""
    nx = c.shape[0]

    # Constant part
    dx = c.copy().flatten()
    
    # Linear part
    dx += L @ x
    
    # Quadratic part
    for i in range(nx):
        dx[i] += x.T @ Q[:,:,i] @ x
    
    return dx

@dataclass
class Model_LosslessQuad:
    """Class for Lossless Quadratic dynamical systems."""
    name: str
    nx: int
    
    # Dynamics at the origin: xdot = c + Lx + phi(x)
    c: np.ndarray
    L: np.ndarray
    Ls: np.ndarray = field(init=False)  # Symmetric part of linear dynamics
    La: np.ndarray = field(init=False)  # Asymmetric part of linear dynamics
    Q: np.ndarray
    
    # Dynamics at the coordinate shift ydot = d + Ay + phi(y)
    m: Union[np.ndarray, cp.Variable, cp.Parameter] = field(init=False)   # Shifted coordinate, default to be 0
    d: Union[np.ndarray, cp.Expression] = field(init=False)   # Constant dynamics
    A: Union[np.ndarray, cp.Expression] = field(init=False)   # Linear dynamics
    As: Union[np.ndarray, cp.Expression] = field(init=False)  # Symmetric part of linear dynamics
    Aa: Union[np.ndarray, cp.Expression] = field(init=False)  # Asymmetric part of linear dynamics
    
    # Function handles
    ode: Callable = field(init=False)
    dK0: Callable = field(init=False)
    ode_shifted: Callable = field(init=False)
    dKm: Callable = field(init=False)
    
    def __init__(self, name: str, c: np.ndarray, L: np.ndarray, Q: np.ndarray, 
                 m: Optional[Union[np.ndarray, cp.Variable, cp.Parameter]] = None):
        """Initialize the model with basic parameters."""
        self.name = name
        self.nx = c.shape[0]
        self.c = c
        self.L = L
        self.Q = Q
        
        # Compute symmetric and asymmetric parts
        self.Ls = 0.5 * (L + L.T)
        self.La = L - self.Ls
        
        # Setup dynamics
        self.setup_dyn()
        
        # Handle coordinate shift
        if m is None:
            m = np.zeros((self.nx, 1))
        self.func_ShiftSystem(m)
    
    def setup_dyn(self):
        """Setup the dynamic functions."""
        self.ode = lambda t, x: ode_quadraticDyn(self.c, self.L, self.Q, t, x)
        self.dK0 = lambda x: self.c.T @ x + x.T @ self.Ls @ x
    
    def func_ShiftSystem(self, m: Union[np.ndarray, cp.Variable, cp.Parameter]):
        """Update the system for a coordinate shift."""
        # Determine if m is a CVXPY object
        is_cvxpy_m = isinstance(m, (cp.Variable, cp.Parameter))

        # Update constant part
        self.m = m
        if is_cvxpy_m:
            d_components = []
            for i in range(self.nx):
                d_i = self.c[i] + self.L[i,:] @ m + cp.quad_form(m, self.Q[:,:,i])
                d_components.append(d_i)
            self.d = cp.vstack(d_components)
        else:
            self.d = np.zeros((self.nx, 1))
            for i in range(self.nx):
                self.d[i] = (self.c[i] +
                            self.L[i,:] @ m +
                            m.T @ self.Q[:,:,i] @ m)

        # Update linear part
        if is_cvxpy_m:
            A_components = []
            for i in range(self.nx):
                A_i = self.L[i,:] + 2 * m.T @ self.Q[:,:,i]
                A_components.append(A_i)
            self.A = cp.vstack(A_components)
        else:
            self.A = np.zeros((self.nx, self.nx))
            for i in range(self.nx):
                self.A[i,:] = self.L[i,:] + 2 * m.T @ self.Q[:,:,i]

        # Update symmetric parts
        if is_cvxpy_m:
            self.As = self.Ls
            for i in range(self.nx):
                self.As = self.As - m[i] * self.Q[:,:,i]
        else:
            self.As = self.Ls.copy()
            for i in range(self.nx):
                self.As = self.As - m[i] * self.Q[:,:,i]
        
        self.Aa = self.A - self.As
        
        # Update ode and power functions
        # These will only work if d, A, Q are numpy arrays
        if not is_cvxpy_m:
            self.ode_shifted = lambda t, x: ode_quadraticDyn(self.d, self.A, self.Q, t, x)
            self.dKm = lambda x: self.d.T @ x + x.T @ self.As @ x
        
        # Create a copy of self
        return self
    
    def get_inverseTimeModel(self):
        """Get the inverse-time model."""
        cInv = -self.c
        LInv = -self.L
        QInv = -self.Q
        
        return Model_LosslessQuad(self.name, cInv, LInv, QInv, self.m) 